src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java

```diff
@@ -28,37 +31,50 @@
 
     private final ParticipationService participationService;
 
-    private final ParticipationRepository participationRepository;
+    private final ProgrammingExerciseParticipationService programmingExerciseParticipationService;
+
+    private final ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository;
 
     private final Optional<VersionControlService> versionControlService;
 
     private final Optional<ContinuousIntegrationService> continuousIntegrationService;
 
     private final SimpMessageSendingOperations messagingTemplate;
 
-    public ProgrammingSubmissionService(ProgrammingSubmissionRepository programmingSubmissionRepository, ParticipationRepository participationRepository,
-            Optional<VersionControlService> versionControlService, Optional<ContinuousIntegrationService> continuousIntegrationService, ParticipationService participationService,
-            SimpMessageSendingOperations messagingTemplate) {
+    public ProgrammingSubmissionService(ProgrammingSubmissionRepository programmingSubmissionRepository,
+            ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository, Optional<VersionControlService> versionControlService,
+            Optional<ContinuousIntegrationService> continuousIntegrationService, ParticipationService participationService, SimpMessageSendingOperations messagingTemplate,
+            ProgrammingExerciseParticipationService programmingExerciseParticipationService) {
         this.programmingSubmissionRepository = programmingSubmissionRepository;
-        this.participationRepository = participationRepository;
+        this.programmingExerciseStudentParticipationRepository = programmingExerciseStudentParticipationRepository;
         this.versionControlService = versionControlService;
         this.continuousIntegrationService = continuousIntegrationService;
         this.participationService = participationService;
         this.messagingTemplate = messagingTemplate;
+        this.programmingExerciseParticipationService = programmingExerciseParticipationService;
     }
 
     public void notifyPush(Long participationId, Object requestBody) {
-        Optional<Participation> optionalParticipation = participationRepository.findById(participationId);
-        if (!optionalParticipation.isPresent()) {
-            log.warn("Invalid participation received while notifying about push: " + participationId);
-            return;
+        ProgrammingExerciseParticipation participation;
+        // We don't know here what kind of participation to expect, so the only way is to try each one.
+        // TODO: This seems like a very bad way to handle this, but is there another way without changing the bitbucket / bamboo configuration?
+        try {
```

Comment:
Here we only get a participationId, so we don't know from which repository to retrieve the participation.
(The implementation with try catch is just for illustration)

Reply:
This can be resolved with the generic ParticipationRepository.